### PR TITLE
Add support for bfalg-ndwi-py

### PR DESCRIPTION
**Pull request by bbarden**

### CODE REVIEW

File: bfapi/service/jobs.py [hunk 1]
```diff
@@ -219,6 +208,26 @@ def create(
         tide_max_24h=tide_max,
     )
 
+def __genCmd(
```
Comment (bbarden) [inline bfapi/service/jobs.py]: renamed to `__make_cli_cmd`.  The `__` prefix is deliberate.  There is no reason for this function to ever be called externally, and scoping is valuable.

File: bfapi/service/jobs.py [hunk 2]
```diff
@@ -219,6 +208,26 @@ def create(
...
+        svc_format: str,
+        geotiff_filenames: []) -> str:
+    log = logging.getLogger(__name__)
+    if svc_format == "pzsvc-ossim":
```
Comment (bbarden) [inline bfapi/service/jobs.py]: renamed to algo_interface

File: bfapi/service/jobs.py [hunk 3]
```diff
@@ -219,6 +208,26 @@ def create(
...
+                    'shoreline',
+                    '--image ' + ','.join(geotiff_filenames),
+                    '--projection geo-scaled',
+                    '--threshold 0.5',
+                    '--tolerance 0.075',
+                    'shoreline.geojson',
+                ])
+    elif svc_format == "pzsvc-ndwi-py":
+        return ' '.join(["--b1", geotiff_filenames[0], "--b2", geotiff_filenames[1], "--fout ./shoreline.json"])
+    elif svc_format == "test-algo-interface":
```
Comment (bbarden) [inline bfapi/service/jobs.py]: removed

File: bfapi/service/jobs.py [hunk 4]
```diff
@@ -219,6 +208,26 @@ def create(
...
+        return "algo-test-command"
+    else:
```
Comment (bbarden) [inline bfapi/service/jobs.py]: `test_throws_when_algorithm_is_unknown()` added

File: bfapi/service/jobs.py [hunk 5]
```diff
@@ -219,6 +208,26 @@ def create(
...
+        log.error('cmd string requested for unknown service type')
+        raise Exception()
```
Comment (bbarden) [inline bfapi/service/jobs.py]: changed to `PreprocessingError`.  explanatory comment added.

File: test/service/test_jobs.py [hunk 6]
```diff
@@ -1359,11 +1351,11 @@ def test_can_handle_large_number_of_cycles(self):
 # Helpers
 #
 
-def create_algorithm():
+def create_algorithm(algo_type='test-algo-interface'):
     return bfapi.service.algorithms.Algorithm(
         bands=('test-algo-band-1', 'test-algo-band-2'),
         description='test-algo-description',
-        interface='test-algo-interface',
+        interface=algo_type,
```
Comment (bbarden) [inline test/service/test_jobs.py]: renamed to `algo_interface`

File: bfapi/service/jobs.py [hunk 7]
```diff
@@ -140,23 +140,12 @@ def create(
     # Dispatch to Piazza
     try:
         log.info('Dispatching <scene:%s> to <algo:%s>', scene_id, algorithm.name)
+        newCmd = __genCmd(algorithm.interface, geotiff_filenames)
```
Comment (bbarden) [inline bfapi/service/jobs.py]: renamed to `cli_cmd`.  More precise.

File: bfapi/service/jobs.py (hunk 1 above)
Comment (dbazile) [inline bfapi/service/jobs.py]: 1. `__` prefix is not a convention used anywhere in the codebase and its use is considered somewhat controversial.
2. This particular usage of `__` prefix actually has no effect:

```
____ [~/code/_experiments/bfalg-ndwi-integration] ____ $ python3
Python 3.5.1 (v3.5.1:37a07cee5969, Dec  5 2015, 21:12:44)
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import bfapi.service.jobs as j
>>> j.__make_cli_cmd('pzsvc-ossim', ['alpha', 'bravo'])
'shoreline --image alpha,bravo --projection geo-scaled --threshold 0.5 --tolerance 0.075 shoreline.geojson'
>>>
```

As such, can you please change it to a single underscore?

File: bfapi/service/jobs.py [hunk 8]
```diff
@@ -208,11 +208,15 @@ def create(
         tide_max_24h=tide_max,
     )
 
-def __genCmd(
-        svc_format: str,
+# The services descended from pzsvc-exec all connect to a CLI (command-line interface) program.
```
Comment (dbazile) [inline bfapi/service/jobs.py]: Two things:

1. The general convention for function documentation is docstrings, ala:

```python
def do_something(x):
    """Take x and multiply it a few times"""
    return x * 3 * 2 * 1
```

2. This comment will go stale eventually as many developers forget to update comments when making code changes.  A descriptive function name should be enough to express what a maintainer needs to know about this function.  I'll defer to industry wisdom on this one:

> “A long descriptive name is better than a short enigmatic name. A long descriptive name is better than a long descriptive comment.” 
> ― Robert C. Martin, Clean Code: A Handbook of Agile Software Craftsmanship

File: bfapi/service/jobs.py [hunk 9]
```diff
@@ -221,13 +225,12 @@ def __genCmd(
...
+    elif algo_interface == "pzsvc-ndwi-py":
         return ' '.join(["--b1", geotiff_filenames[0], "--b2", geotiff_filenames[1], "--fout ./shoreline.json"])
-    elif svc_format == "test-algo-interface":
-        return "algo-test-command"
     else:
-        log.error('cmd string requested for unknown service type')
-        raise Exception()
+        error_string = 'CLI cmd string requested for unknown algorithm interface "' + algo_interface + '".'
+        log.error(error_string)
+        raise PreprocessingError(message=error_string)
```
Comment (dbazile) [inline bfapi/service/jobs.py]: Avoid using data types in variable names, ala `error_string`.  `error_message` is more conventional and conveys more meaning.

I also should have mentioned this earlier, but `PreprocessingError` gets serialized and sent with the HTTP response so it needs to be cleaned up.  This message can probably be shortened to `unknown algorithm interface 'foo'`.

File: test/service/test_jobs.py [hunk 10]
```diff
@@ -73,46 +73,46 @@ def create_mock(self, target_name):
         return patcher.start()
 
     def test_returns_job(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 11]
```diff
@@ -73,46 +73,46 @@ def create_mock(self, target_name):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
 
     def test_assigns_correct_algorithm_name(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 12]
```diff
@@ -73,46 +73,46 @@ def create_mock(self, target_name):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
         self.assertEqual('test-algo-name', job.algorithm_name)
 
     def test_assigns_correct_algorithm_version(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 13]
```diff
@@ -73,46 +73,46 @@ def create_mock(self, target_name):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
         self.assertEqual('test-algo-version', job.algorithm_version)
 
     def test_assigns_correct_created_by(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 14]
```diff
@@ -73,46 +73,46 @@ def create_mock(self, target_name):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
         self.assertEqual('test-user-id', job.created_by)
 
     def test_assigns_correct_created_on(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 15]
```diff
@@ -73,46 +73,46 @@ def create_mock(self, target_name):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
         self.assertEqual(datetime.utcnow().date(), job.created_on.date())
 
     def test_assigns_correct_geometry(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 16]
```diff
@@ -121,7 +121,7 @@ def test_assigns_correct_geometry(self):
                          job.geometry)
 
     def test_assigns_correct_job_id(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 17]
```diff
@@ -130,47 +130,47 @@ def test_assigns_correct_job_id(self):
         self.assertEqual('test-new-job-id', job.job_id)
 
     def test_assigns_correct_name(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 18]
```diff
@@ -130,47 +130,47 @@ def test_assigns_correct_job_id(self):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
         self.assertEqual('test-name', job.name)
 
     def test_assigns_correct_scene_capture_date(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 19]
```diff
@@ -130,47 +130,47 @@ def test_assigns_correct_job_id(self):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
         self.assertEqual('2014-05-13T16:53:20', job.scene_capture_date.isoformat())
 
     def test_assigns_correct_scene_sensor_name(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 20]
```diff
@@ -130,47 +130,47 @@ def test_assigns_correct_job_id(self):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
         self.assertEqual('test-sensor-name', job.scene_sensor_name)
 
     def test_assigns_correct_scene_id(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 21]
```diff
@@ -130,47 +130,47 @@ def test_assigns_correct_job_id(self):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
         self.assertEqual('test-scene-id', job.scene_id)
 
     def test_assigns_correct_status(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 22]
```diff
@@ -130,47 +130,47 @@ def test_assigns_correct_job_id(self):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         job = jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-service-id', 'test-name')
         self.assertIsInstance(job, jobs.Job)
         self.assertEqual('Submitted', job.status)
 
     def test_requests_tide_prediction(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 23]
```diff
@@ -184,7 +184,7 @@ def test_requests_tide_prediction(self):
         }, self.mock_requests.request_history[0].json())
 
     def test_fetches_correct_scene(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 24]
```diff
@@ -226,7 +226,7 @@ def test_sends_correct_payload_to_piazza_pzsvc_ndwi_py(self):
         }, json.loads(self.mock_execute.call_args[0][2]['body']['content']))
 
     def test_does_not_create_database_record_if_cannot_start(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 25]
```diff
@@ -238,7 +238,7 @@ def test_does_not_create_database_record_if_cannot_start(self):
         self.assertEqual(0, self.mock_insert_job_user.call_count)
 
     def test_logs_creation_success(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 26]
```diff
@@ -258,7 +258,7 @@ def test_logs_creation_failure_during_algorithm_retrieval(self):
         ], logstream.getvalue().splitlines())
 
     def test_logs_creation_failure_during_scene_retrieval(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 27]
```diff
@@ -269,7 +269,7 @@ def test_logs_creation_failure_during_scene_retrieval(self):
         ], logstream.getvalue().splitlines())
 
     def test_logs_creation_failure_during_tide_prediction(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 28]
```diff
@@ -280,7 +280,7 @@ def test_logs_creation_failure_during_tide_prediction(self):
         ], logstream.getvalue().splitlines())
 
     def test_discards_malformed_tide_service_response(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 29]
```diff
@@ -304,7 +304,7 @@ def test_discards_malformed_tide_service_response(self):
         ], logstream.getvalue().splitlines())
 
     def test_logs_creation_failure_during_execution(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 30]
```diff
@@ -317,7 +317,7 @@ def test_logs_creation_failure_during_execution(self):
         ], logstream.getvalue().splitlines())
 
     def test_logs_creation_failure_during_database_insert(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 31]
```diff
@@ -330,15 +330,15 @@ def test_logs_creation_failure_during_database_insert(self):
         ], logstream.getvalue().splitlines())
 
     def test_throws_when_piazza_throws(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 32]
```diff
@@ -330,15 +330,15 @@ def test_logs_creation_failure_during_database_insert(self):
...
         self.mock_get_scene.return_value = create_scene()
         self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         self.mock_execute.side_effect = piazza.ServerError(500)
         with self.assertRaises(piazza.ServerError):
             jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-algo-id', 'test-name')
 
     def test_throws_when_database_insertion_fails(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 33]
```diff
@@ -347,17 +347,19 @@ def test_throws_when_database_insertion_fails(self):
...
     def test_throws_when_algorithm_not_found(self):
         self.mock_get_algo.side_effect = bfapi.service.algorithms.NotFound('test-algo-id')
+        self.mock_get_scene.return_value = create_scene()
+        self.mock_requests.post('/tides', text=RESPONSE_TIDE)
         with self.assertRaises(jobs.PreprocessingError):
             jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-algo-id', 'test-name')
 
     def test_throws_when_scene_not_found(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 34]
```diff
@@ -347,17 +347,19 @@ def test_throws_when_database_insertion_fails(self):
...
         self.mock_get_scene.side_effect = bfapi.service.scenes.NotFound('test-scene-id')
         with self.assertRaises(jobs.PreprocessingError):
             jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-algo-id', 'test-name')
 
     def test_throws_when_scene_is_missing_required_bands(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: test/service/test_jobs.py [hunk 35]
```diff
@@ -366,13 +368,20 @@ def test_throws_when_scene_is_missing_required_bands(self):
             jobs.create(API_KEY, 'test-user-id', 'test-scene-id', 'test-algo-id', 'test-name')
 
     def test_throws_when_catalog_is_unreachable(self):
-        self.mock_get_algo.return_value = create_algorithm()
+        self.mock_get_algo.return_value = create_algorithm('pzsvc-ndwi-py')
```
Comment (dbazile) [inline test/service/test_jobs.py]: `pzsvc-ndwi-py` is the default value for the `create_algorithm` function.  You don't need to repeat it here.

File: bfapi/service/jobs.py (hunk 8 above)
Comment (bbarden) [inline bfapi/service/jobs.py]: 1. Good to know
2. This function is explicitly for the purpose of creating cmd strings to fit this particular case.  If someone is making a change to this function that causes the comment to become inaccurate, that change itself is a mistake.
2a. "industry wisdom" is in no way unanimous on this subject.

Comment (dbazile) [inline bfapi/service/jobs.py]: > This function is explicitly for the purpose of creating cmd strings to fit this particular case. If someone is making a change to this function that causes the comment to become inaccurate, that change itself is a mistake.

Any "safety" provided by this comment is an illusion.  The truth is, developers _will_ forget to update things like this when the underlying assumptions change.  With that in mind, there is a significantly higher probability that the function name will be changed before the comment will.

[Here's another perspective on this](http://softwareengineering.stackexchange.com/questions/254978/whats-wrong-with-comments-that-explain-complex-code).

File: bfapi/service/jobs.py [hunk 36]
```diff
@@ -219,6 +208,26 @@ def create(
         tide_max_24h=tide_max,
     )
 
+def __make_cli_cmd(
```
Comment (dbazile) [inline bfapi/service/jobs.py]: Can you replace the `__` prefix with `_`?

Comment (bbarden) [inline bfapi/service/jobs.py]: That would make it not private.  It is intended to be a private function.

Comment (bbarden) [inline bfapi/service/jobs.py]: If you require it, though, I will do it.

Comment (bbarden) [inline bfapi/service/jobs.py]: name changed to `__make_algorithm_cli_cmd`

File: bfapi/service/jobs.py [hunk 37]
```diff
@@ -208,7 +208,7 @@ def create(
         tide_max_24h=tide_max,
     )
 
-def __make_cli_cmd(
+def __make_algorithm_cli_cmd(
```
Comment (dbazile) [inline bfapi/service/jobs.py]: See, now you're just messing with my head. :trollface: 

Can you please describe to me what you think happens when you prefix this particular function with `__`?

Comment (bbarden) [inline bfapi/service/jobs.py]: My understanding is that it effectively makes it a private function - that it prevents the function from being accessed outside of the local area.  (I don't recall if the "local area" refers to the file itself or the namespace or whatever, but that's the basic idea.  I spent more focus/research on it when I was making the decision).  If I'm incorrect on that, then you have my apologies.
It's notably distinct from naming a function `__blah__` though.  I do recall that.

Comment (dbazile) [inline bfapi/service/jobs.py]: It doesn't work that way for module-level functions, Ben.  Please check this gist:

https://gist.github.com/dbazile/c3a0eac7963ee36554c9961f0eae2f39

Comment (bbarden) [inline bfapi/service/jobs.py]: In general, I prefer to make functions private unless there's some specific reason to expose them.  On code-reading, it makes it clear that that particular function is not being used elsewhere, and it cuts down on the possibility of improper external use by people who don't fully understand the context of the function.  I have had to spend too much time dealing with the code of people who only thought that their code was self-explanatory.  Context helps.

**Pull request merged**